client/comms/wsconn.go

```diff
@@ -0,0 +1,320 @@
+}
+
+// isConnected returns the connection connected state.
+func (conn *WsConn) isConnected() bool {
+	conn.connectedMtx.RLock()
+	defer conn.connectedMtx.RUnlock()
+	return conn.connected
+}
+
+// setConnected updates the connection's connected state.
+func (conn *WsConn) setConnected(connected bool) {
+	conn.connectedMtx.Lock()
+	conn.connected = connected
+	conn.connectedMtx.Unlock()
+}
+
+// logRoute logs a request keyed by its id.
+func (conn *WsConn) logRequest(id uint64, req *msgjson.Message) {
+	conn.reqMtx.Lock()
+	conn.req[id] = req
+	conn.reqMtx.Unlock()
+}
+
+// FetchRequest fetches the request associated with the id. The returned
+// request is removed from the cache.
+func (conn *WsConn) FetchRequest(id uint64) (*msgjson.Message, error) {
+	conn.reqMtx.Lock()
+	defer conn.reqMtx.Unlock()
+	req := conn.req[id]
+	if req == nil {
+		return nil, fmt.Errorf("no request found for id %d", id)
+	}
+
+	delete(conn.req, id)
+
+	return req, nil
+}
+
+// NextID returns the next request id.
+func (conn *WsConn) NextID() uint64 {
+	return atomic.AddUint64(&conn.rID, 1)
+}
+
+// close terminates all websocket processes and closes the connection.
+func (conn *WsConn) close() {
+	conn.wsMtx.Lock()
+	defer conn.wsMtx.Unlock()
+	msg := websocket.FormatCloseMessage(websocket.CloseNormalClosure, "")
+	conn.ws.WriteControl(websocket.CloseMessage, msg,
+		time.Now().Add(conn.cfg.WriteWait))
+	conn.ws.Close()
+}
+
+// connect attempts to establish a websocket connection.
+func (conn *WsConn) connect() error {
+	url := url.URL{
+		Scheme: "wss",
+		Host:   conn.cfg.Host,
+		Path:   conn.cfg.Path,
+	}
+
+	dialer := &websocket.Dialer{
+		Proxy:            http.ProxyFromEnvironment,
+		HandshakeTimeout: 10 * time.Second,
+		TLSClientConfig:  conn.tlsCfg,
+	}
+
+	ws, _, err := dialer.Dial(url.String(), nil)
+	if err != nil {
+		return err
+	}
+
+	ws.SetPingHandler(func(string) error {
+		now := time.Now()
+		err := ws.SetReadDeadline(now.Add(conn.cfg.PingWait))
+		if err != nil {
+			log.Errorf("read deadline error: %v", err)
+			return err
+		}
+
+		// Respond with a pong.
+		err = ws.WriteControl(websocket.PongMessage, []byte{},
+			now.Add(conn.cfg.WriteWait))
+		if err != nil {
+			log.Errorf("pong error: %v", err)
+			return err
+		}
+
+		return nil
+	})
+
+	conn.wsMtx.Lock()
+	conn.ws = ws
+	conn.wsMtx.Unlock()
+
+	return nil
+}
+
+// read fetches and parses incoming messages for processing. This should be
+// run as a goroutine.
+func (conn *WsConn) read() {
+	for {
+		msg := new(msgjson.Message)
+
+		conn.wsMtx.Lock()
+		ws := conn.ws
+		conn.wsMtx.Unlock()
+
+		err := ws.ReadJSON(msg)
+		if err != nil {
+			if _, ok := err.(*json.UnmarshalTypeError); ok {
+				// JSON decode errors are not fatal, log and proceed.
+				log.Errorf("json decode error: %v", err)
+				continue
+			}
+
+			if websocket.IsCloseError(err, websocket.CloseGoingAway,
+				websocket.CloseNormalClosure) ||
+				strings.Contains(err.Error(), "websocket: close sent") {
```

Comment:
I'm still confused about this.  The remote host sends the close code, so the client is at the mercy of the remote host.  Why not catch any and all close errors regardless of the code (`IsUnexpectedCloseError`)?  This is obviously much more important on the server side to prevent abuse from clients however.

Reply:
I'm discerning context from the close error returned, my thinking is if a normal close error is returned then a shutdown is intended so terminate immediately and don't reconnect. if it's a read error using a closed connection then that also means a shutdown was intended, all other read errors will then be as a result of a dropped connection, for those cases we try and reconnect.

This is why I'm handling specific error cases for a shutdown before handling all others since those trigger a reconnection.